pkg/kubectl/util/service_port.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+	api "k8s.io/kubernetes/pkg/apis/core"
+)
+
+// Lookup containerPort number by its named port name
+func LookupContainerPortNumberByName(pod api.Pod, name string) (int32, error) {
+	matchfound := false
+	namedport := int32(-1)
+	for _, ctr := range pod.Spec.Containers {
+		for _, ctrportspec := range ctr.Ports {
+			if ctrportspec.Name == name {
+				namedport = ctrportspec.ContainerPort
+				matchfound = true
+			}
+		}
+	}
+
+	if matchfound == false {
+		return namedport, fmt.Errorf("Pod '%s' does not have a named port '%s'", pod.Name, name)
+	} else {
+		return namedport, nil
+	}
+}
+
+// Gather service port to target pod mapping from a Service
+// It implements the handling of resolving container named port, as well as ignoring targetPort when clusterIP=None
+func GatherServicePortMapping(svc api.Service, pod api.Pod) (map[int32]int32, error) {
+	mapping := make(map[int32]int32)
+	for _, svcportspec := range svc.Spec.Ports {
+		if svc.Spec.ClusterIP == "None" {
+			mapping[svcportspec.Port] = svcportspec.Port
+		} else {
+			if svcportspec.TargetPort.Type == intstr.Int {
+				if svcportspec.TargetPort.IntValue() == 0 {
+					// targetPort is omitted, and the IntValue() would be zero
+					mapping[svcportspec.Port] = svcportspec.Port
+				} else {
+					mapping[svcportspec.Port] = int32(svcportspec.TargetPort.IntValue())
+				}
+			} else {
+				portnum, err := LookupContainerPortNumberByName(pod, svcportspec.TargetPort.String())
```

Comment:
this has the potential to return an error and short-circuit on a port that we're not even going to use... you could just omit the target from the mapping (or map to zero), and handle missing cases in translateServicePortToTargetPort if we actually try to make use of ports that don't have corresponding ports in the pod

Reply:
Would it be better if we just look up as we process the arguments, and the error would be much clear?  I did the gathering first because it was easier for me to implement at the time.

Reply:
Probably so

Reply:
Realized that lookup failure can't be treated as failures, because service/pod is not required to declare every port a pod listens on, so the point of surfacing clear error is irrelevant.

Reply:
Actually, I think it would work by surfacing only the named port lookup error.

Reply:
services are required to declare the ports they surface, and pods must declare a named port for named mappings to work

> I think it would work by surfacing only the named port lookup error

that sounds right

Reply:
> services are required to declare the ports they surface

So in that expectation, if a Service does not declare a port, but the user still wants to port-forward to that port, should that be treated as error?  Right now I am pass it as-is to match the use case for other types of resources.